### PR TITLE
Jg/checkmate

### PR DESCRIPTION
**Pull request by jeffgerlach**
Implements `checkmate?(king)` method in the game model which takes in a king and determines if it's in checkmate (returns true), otherwise returns false. Not the most elegant of solutions (wanted to beat the time crunch!), but hopefully it's straightforward. 

Needs some refactoring, but I tried to put a ton of comments to help people follow along with the logic. Had to slightly modify @nicolaguidi's `determine_check` method to get it to play along with my method, though the functionality should be very similar. 

Added a bunch of tests for checkmate, along with tests for the helper method which grabs the squares between an enemy piece and the king to see if a friendly piece can block check.

Will need to implement this method into the game/piece controller as part of the turn logic and to actually end the game when a player is put into checkmate. I added a Trello card with that task so this PR doesn't get any bigger than it currently is.

### CODE REVIEW

File: app/controllers/pieces_controller.rb [hunk 1]
```diff
@@ -28,12 +28,6 @@ def update
 
   private
 
-  def flash_notice
```
Comment (jeffgerlach) [inline app/controllers/pieces_controller.rb:37]: this was duplicated from an earlier merge

File: app/models/game.rb [hunk 2]
```diff
@@ -69,21 +69,227 @@ def player_missing?
   end
 
   def determine_check(king)
-    opponents_pieces = pieces.where(color: !king.color)
+    # if king is in check, returns list of pieces that threaten it - will evaluate to true if present
+    # instead of just returning true/false, I have this method return the list of enemy pieces putting the
+    # king in check to try to DRY up the code
+    threatening_pieces = []
+    opponents_pieces = self.pieces.where(color: !king.color, captured: false)
```
Comment (jeffgerlach) [inline app/models/game.rb:76]: added a check to see if pieces are captured to narrow down the piece list and save some processing time

File: app/models/king.rb [hunk 3]
```diff
@@ -59,19 +64,25 @@ def initial_state?(rook)
 
   def puts_in_check?(rook)
     if queenside?(rook) # rook on queenside
-      first_y_space = -1 # move left one
-      second_y_space = -2 # move left two
+      first_x_space = -1 # move left one
+      #second_x_space = -2 # move left two
```
Comment (jeffgerlach) [inline app/models/king.rb:68]: also fixed up the castling method on the backend to actually implement check correctly

File: app/models/rook.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class Rook < Piece
   def valid_move?(dest_x, dest_y)
-    dest_piece = game.pieces
+    dest_piece = self.game.pieces
```
Comment (jeffgerlach) [inline app/models/rook.rb]: not necessary, forgot I put that in there while testing....won't hurt either i suppose

File: spec/models/game_spec.rb [hunk 5]
```diff
@@ -297,4 +297,271 @@
       expect(game.check?).to eq(true)
     end
   end
+
+  describe "range_between_pieces" do
+    before :each do
```
Comment (jeffgerlach) [inline spec/models/game_spec.rb:302]: helper method tests

File: spec/models/king_spec.rb [hunk 6]
```diff
@@ -145,18 +149,23 @@
         expect(@white_king.can_castle?(white_queenside_rook2)).to be false
       end
       context "start, moves through, or ends in check" do
-        # check is not currently implemented - will need to fill these in
         it "does not allow castle if the king currently is in check" do
           # set up situation where king is in check to start
-          expect(false).to be false
+          @black_kingside_rook.update_attributes(x_position: 4, y_position: 5)
+          expect(@white_king.can_castle?(@white_kingside_rook)).to be false
+          expect(@white_king.flash_message).to eq "Can't castle while in check!"
```
Comment (jeffgerlach) [inline spec/models/king_spec.rb:156]: updated some castling checks since check method is implemented

File: app/models/game.rb (hunk 2 above)
Comment (nicolaguidi) [inline app/models/game.rb:76]: Nice!

**Pull request merged**